src/tree/updater_quantile_hist.cc

```diff
@@ -574,15 +574,55 @@ void QuantileHistMaker::Builder::InitData(const GHistIndexMatrix& gmat,
       CHECK_EQ(param_.sampling_method, TrainParam::kUniform)
         << "Only uniform sampling is supported, "
         << "gradient-based sampling is only support by GPU Hist.";
-      std::bernoulli_distribution coin_flip(param_.subsample);
       auto& rnd = common::GlobalRandom();
+#if XGBOOST_CUSTOMIZE_GLOBAL_PRNG
```

Comment:
Could you, please, explain the reason to add the define?

Reply:
Introduced optimizations are applicable only  for std rng engines, but if `XGBOOST_CUSTOMIZE_GLOBAL_PRNG` was defined internal `CustomGlobalRandomEngine` is used (see https://github.com/dmlc/xgboost/blob/master/src/common/random.h#L37).
`CustomGlobalRandomEngine` doesn't have implementation of `discard` method.